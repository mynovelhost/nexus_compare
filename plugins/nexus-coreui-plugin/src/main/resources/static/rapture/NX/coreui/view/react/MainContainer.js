/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
/*global Ext, NX, REACT_COMPONENTS, react, ReactDOM */

/**
 * React Main Container
 *
 * @since 3.21
 */
Ext.define('NX.coreui.view.react.MainContainer', {
  extend: 'Ext.Component',
  alias: 'widget.nx-coreui-react-main-container',

  reactView: undefined,

  scrollable: true,

  listeners: {
    afterrender: 'initReactView',
    beforedestroy: 'destroyReactView'
  },

  initReactView: function() {
    var reactElement = react.createElement(this.reactView);
    ReactDOM.render(reactElement, this.getEl().dom);
  },

  destroyReactView: function() {
    if (this.reactView) {
      ReactDOM.unmountComponentAtNode(this.getEl().dom);
    }
  },

  refresh: function() {
    this.destroyReactView();
    this.initReactView();
  }
});
