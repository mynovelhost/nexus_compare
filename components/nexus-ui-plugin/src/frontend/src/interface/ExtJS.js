/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
/*global Ext, NX*/

import {useEffect, useState} from 'react';

/**
 * @since 3.22
 */
export default class {
  /**
   * Open a success message notification
   * @param text
   */
  static showSuccessMessage(text) {
    NX.Messages.success(text)
  }

  /**
   * Open an error message notification
   * @param text
   */
  static showErrorMessage(text) {
    NX.Messages.error(text);
  }

  /**
   *@returns a complete url for the current nexus instance
   */
  static urlOf(path) {
    return NX.util.Url.urlOf(path);
  }

  /**
   * Set the global dirty status to prevent accidental navigation
   * @param key - a unique key for the view that is dirty
   * @param isDirty - whether the view is dirty or not
   */
  static setDirtyStatus(key, isDirty) {
    window.dirty = window.dirty || [];

    if (isDirty && window.dirty.indexOf(key) === -1) {
      window.dirty.push(key);
    }
    else {
      window.dirty = window.dirty.filter(it => it !== key);
    }
  }

  /**
   * @return {location: {pathname: string }}
   */
  static useHistory({basePath}) {
    const [path, setPath] = useState(Ext.History.getToken());

    useEffect(() => {
      Ext.History.on('change', setPath);
      return () => Ext.History.un('change', setPath);
    });

    return {
      location: {
        pathname: path.replace(basePath, '')
      }
    };
  }

  /**
   * Set the breadcrumbs to show. If the items argument is not provided we'll show the "root" breadcrumb.
   * @param items [{itemName}]
   */
  static setBreadcrumbs(items = []) {
    const content = Ext.ComponentQuery.query('#feature-content')[0];
    const breadcrumb = content.down('#breadcrumb');

    if (!items.length) {
      // If we're on the root page we should show the root breadcrumbs
      content.showRoot();
    }
    else {
      // If we have items, then we need to display breadcrumbs to allow the user to go back
      const breadcrumbs = [
        {
          xtype: 'container',
          itemId: 'nx-feature-icon',
          width: 32,
          height: 32,
          cls: content.currentIcon,
          ariaRole: 'presentation'
        },
        {
          xtype: 'button',
          itemId: 'nx-feature-name',
          scale: 'large',
          ui: 'nx-drilldown',
          text: content.currentTitle,
          handler: function() {
            const menuController = Ext.getApplication().getController('NX.controller.Menu');
            const path = menuController.currentSelectedPath.slice(1).toLowerCase();
            const currentPath = Ext.History.getToken();
            if (path === currentPath) {
              const refreshController = Ext.getApplication().getController('NX.controller.Refresh');
              refreshController.refresh();
            }
            else {
              Ext.History.add(path);
            }
          }
        }
      ];
      items.forEach(item => {
        if (item.itemName) {
          breadcrumbs.push({
            xtype: 'label',
            cls: 'nx-breadcrumb-separator',
            text: '/',
            ariaRole: 'presentation',
            tabIndex: -1
          });

          breadcrumbs.push({
            xtype: 'button',
            scale: 'medium',
            ui: 'nx-drilldown',
            text: Ext.htmlEncode(item.itemName),
            disabled: !item.itemPath,
            handler: function() {
              Ext.History.add(item.itemPath);
            }
          })
        }
      });

      breadcrumb.removeAll();
      breadcrumb.add(breadcrumbs);
    }
  }

  static requestConfirmation({title, message, yesButtonText = 'Yes', noButtonText = 'No'}) {
    const options = {
      buttonText: {
        yes: yesButtonText,
        no: noButtonText
      }
    };

    return new Promise((resolve, reject) => NX.Dialogs.askConfirmation(title, message, resolve, {
      ...options,
      onNoFn: reject
    }));
  }

  /**
   * Create a Promise that will fetch an authentication token using the
   * username and password supplied
   * @param username
   * @param password
   * @returns {Promise}
   */
  static fetchAuthenticationToken(username, password) {
    const b64u = NX.util.Base64.encode(username);
    const b64p = NX.util.Base64.encode(password);
    return new Promise((resolve, reject) => {
      NX.direct.rapture_Security.authenticationToken(b64u, b64p, resolve);
    });
  }

  /**
   * Prompt the user to re-authenticate to fetch an authentication token
   * @param message prompt shown to user
   * @returns {Promise}
   */
  static requestAuthenticationToken(message) {
    return new Promise((resolve, reject) => {
      NX.Security.doWithAuthenticationToken(
          message,
          {
            success: function(authToken) {
              return resolve(authToken);
            },
            failure: function() {
              return reject();
            }
          }
      );
    });
  }

  static downloadUrl(url) {
    NX.util.DownloadHelper.downloadUrl(url);
  }

  static state() {
    return NX.State;
  }
}
