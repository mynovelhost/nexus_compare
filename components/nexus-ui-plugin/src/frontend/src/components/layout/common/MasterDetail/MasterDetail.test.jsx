/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import React from 'react';
import {fireEvent, render} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import MasterDetail from './MasterDetail';
import Master from './Master';
import Detail from './Detail';

import ExtJS from '../../../../interface/ExtJS';

jest.mock('../../../../interface/ExtJS', () => class {
  static useHistory = jest.fn();
  static setBreadcrumbs = jest.fn();
});

describe('MasterDetail', () => {
  it('renders the master view when at the root', () => {
    ExtJS.useHistory.mockReturnValue({location: {pathname: ''}});
    const {container, queryByTestId} = render(
        <MasterDetail path="admin">
          <Master><TestMaster /></Master>
          <Detail><TestDetail /></Detail>
        </MasterDetail>
    );

    expect(queryByTestId('master')).toBeInTheDocument();
    expect(queryByTestId('detail')).not.toBeInTheDocument();
    expect(ExtJS.setBreadcrumbs).toBeCalledWith();
    expect(container).toMatchSnapshot();
  });

  it('renders the detail view when on a child path', () => {
    ExtJS.useHistory.mockReturnValue({location: {pathname: ':itemId'}});
    const {container, queryByTestId} = render(
        <MasterDetail path="admin">
          <Master><TestMaster /></Master>
          <Detail editBreadcrumb={(itemId) => `Edit ${itemId}`}><TestDetail /></Detail>
        </MasterDetail>
    );

    expect(queryByTestId('master')).not.toBeInTheDocument();
    expect(queryByTestId('detail')).toBeInTheDocument();
    expect(ExtJS.setBreadcrumbs).toBeCalledWith([{itemName: 'Edit itemId'}]);
    expect(container).toMatchSnapshot();
  });

  it('renders the detail view when onCreate is called', () => {
    ExtJS.useHistory.mockReturnValue({location: {pathname: ''}});
    const {container, getByTestId, queryByTestId} = render(
        <MasterDetail path="admin">
          <Master><TestMaster /></Master>
          <Detail createBreadcrumb="Create"><TestDetail /></Detail>
        </MasterDetail>
    );

    expect(queryByTestId('master')).toBeInTheDocument();
    expect(ExtJS.setBreadcrumbs).toBeCalledWith();
    fireEvent.click(getByTestId('create'));

    expect(queryByTestId('master')).not.toBeInTheDocument();
    expect(queryByTestId('detail')).toBeInTheDocument();
    expect(ExtJS.setBreadcrumbs).toBeCalledWith([{itemName: 'Create'}]);
    expect(container).toMatchSnapshot();
  });
});

// These components remove the warnings about the Master/Detail props not being used and provide hooks for validation
function TestMaster({onCreate, onEdit}) {
  return <div data-testid="master"><button data-testid="create" onClick={onCreate} /></div>;
}

function TestDetail({itemId, onDone}) {
  return <div data-testid="detail">${itemId}</div>;
}
