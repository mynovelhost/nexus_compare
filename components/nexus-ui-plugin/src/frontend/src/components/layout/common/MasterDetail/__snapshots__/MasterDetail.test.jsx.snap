// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MasterDetail renders the detail view when on a child path 1`] = `
<div>
  <div
    class="nxrm-master-detail"
  >
    <div
      data-testid="detail"
    >
      $
      itemId
    </div>
  </div>
</div>
`;

exports[`MasterDetail renders the detail view when onCreate is called 1`] = `
<div>
  <div
    class="nxrm-master-detail"
  >
    <div
      data-testid="detail"
    >
      $
      
    </div>
  </div>
</div>
`;

exports[`MasterDetail renders the master view when at the root 1`] = `
<div>
  <div
    class="nxrm-master-detail"
  >
    <div
      data-testid="master"
    >
      <button
        data-testid="create"
      />
    </div>
  </div>
</div>
`;
